infer/src/pulse/PulseExecutionDomain.ml

```diff
@@ -102,6 +108,170 @@ let pp fmt exec_state =
   F.fprintf fmt "%a%a" (pp_header TEXT) exec_state AbductiveDomain.pp (to_astate exec_state)
 
 
+(* Infinite loop detector creates a new InfiniteProgram state when detected *)
+
+let widenstate = AnalysisGlobalState.make_dls ~init:(fun () -> Some (Stdlib.Hashtbl.create 16))
+
+let get_widenstate () = DLS.get widenstate
+
+let add_widenstate key =
+  Utils.with_dls widenstate ~f:(fun widenstate ->
+      match widenstate with
+      | None ->
+          widenstate
+      | Some state ->
+          Stdlib.Hashtbl.add state key () ;
+          widenstate )
+
+
+let rec find_infinite_state (lst : t list) : bool =
+  match lst with
+  | hd :: tl -> (
+    match hd with InfiniteProgram _ -> true | _ -> find_infinite_state tl )
+  | _ ->
+      false
+
+
+let back_edge (prev : t list) (next : t list) (num_iters : int) : t list * int =
+  let has_infinite_state = find_infinite_state next in
+  if has_infinite_state then ([], -1)
+  else
+    let _ = num_iters in
+    let cfgnode = AnalysisState.get_node_exn () in
+    let same = phys_equal prev next in
+    let rec detect_elem e lst curi : bool * int =
+      match lst with
+      | [] ->
+          (false, -1)
+      | hd :: tl ->
+          if leq ~lhs:hd ~rhs:e then (true, curi) else detect_elem e tl (curi + 1)
+    in
+    let rec compute_workset prev next newset nextit : (t * int) list =
+      match next with
+      | [] ->
+          newset
+      | hd :: tl ->
+          let res, idx = detect_elem hd prev 0 in
+          let recidx = match idx with -1 -> nextit | _ -> idx in
+          if res then compute_workset prev tl newset (nextit + 1)
+          else [(hd, recidx)] @ compute_workset prev tl newset (nextit + 1)
+    in
+    let workset = compute_workset prev next [] 0 in
+    let prevlen = List.length prev in
+    let nextlen = List.length next in
+    let worklen = List.length workset in
+    let extract_pathcond hd : Formula.t =
+      match hd with
+      | AbortProgram summary ->
+          AbductiveDomain.Summary.get_path_condition summary
+      | ExitProgram summary ->
+          AbductiveDomain.Summary.get_path_condition summary
+      | LatentAbortProgram a ->
+          AbductiveDomain.Summary.get_path_condition a.astate
+      | LatentInvalidAccess a ->
+          AbductiveDomain.Summary.get_path_condition a.astate
+      | InfiniteProgram astate ->
+          AbductiveDomain.get_path_condition astate
+      | ExceptionRaised astate ->
+          AbductiveDomain.get_path_condition astate
+      | ContinueProgram astate ->
+          AbductiveDomain.get_path_condition astate
+      | LatentSpecializedTypeIssue {astate; _} ->
+          AbductiveDomain.Summary.get_path_condition astate
```

Comment:
This can be simplified a bit. You could also add this as `PulseExecutionDomain.get_path_condition`!
```suggestion
      | AbortProgram astate
      | ExitProgram astate
      | LatentAbortProgram {astate}
      | LatentInvalidAccess {astate}
      | LatentSpecializedTypeIssue {astate} ->
          AbductiveDomain.Summary.get_path_condition summary
      | InfiniteProgram astate
      | ExceptionRaised astate
      | ContinueProgram astate ->
          AbductiveDomain.get_path_condition astate
```